### DISCUSSION TITLE
Cross-Compiling for Windows ARM64?

### DISCUSSION THREAD

**Discussion by andrjohns** [r-universe-org/help#421] [Ideas]
Given the fantastic existing work that went into developing MacOS ARM64 cross-compiling, would any of that translate to cross-compiling binaries for Windows ARM64?

Given that the LLVM toolchain is itself cross-compiled for an aarch64 host and aarch64 target, it could alternatively be cross-compiled for an x86_64 host with an aarch64 target - so that a similar process as the current Macos approach could be used.

**Comment by jeroen** [comment]
I think there is some work under way by @kalibera to experiment with a cross compiler but in general it is often impractical to cross compile R packages. This is because during several steps of the build process, R has to load and run some of the binaries it just built, and of course this does not work for binaries that were built for a target that cannot run on the host platform.

Either way it is not up to me but up to the maintainer of rtools (Tomas Kalibera) to figure out if this is something that R could support.

**Reply by andrjohns** [reply]
Makes sense to me, thanks for clarifying!

**Reply by kalibera** [reply]
Rtools includes cross-compilers for the target platforms (gcc for Windows/x86_64, llvm for Windows/aarch64, compilation running on Linux/x86_64). R has limited support for cross-compiling R itself and of R packages, see file src/gnuwin32/README.compilation in the R source tree. Base R and recommended packages can be cross-compiled (and installer for Windows/aarch64 can be cross-compiled on Windows/x86_64), but many contributed CRAN packages probably won't be cross-compilable, this is not tested and CRAN package authors are not required to support cross-compilation.

**Reply by andrjohns** [reply]
Turns out that LLVM makes the cross-compiling process surprisingly painless. Here's a minimal proof-of-concept for using github actions to cross-compile for Windows ARM64: [cross-test.yml](https://github.com/andrjohns/cross-tests/blob/main/.github/workflows/cross-test.yml), with an example output in the latest [run artifacts](https://github.com/andrjohns/cross-tests/actions/runs/9375376859)

It uses the x86_64 LLVM from MSYS2 to build/link against the aarch64 rtools toolchain and R binaries. The only issue at the moment is that the packaged `.dll` is installed into `libs/x64/` instead of just `libs/`, but other than that the cross-compiled package appears to work without issue

**Comment by yutannihilation** [comment]
FYI, GitHub already announced their plan to support ARM Windows GitHub Actions  runner. So, I think we can just wait. As GitHub is owned by Microsoft, I guess they are not very slow at things related to Windows.

> GitHub hosted runners that is available in private beta with GA expected later this summer  
https://techcommunity.microsoft.com/t5/azure-compute-blog/announcing-the-preview-of-new-azure-vms-based-on-the-azure/ba-p/4146353

**Reply by andrjohns** [reply]
Looks like they're already here: https://github.blog/2024-06-03-arm64-on-github-actions-powering-faster-more-efficient-build-systems/

Only for enterprise/teams accounts at the moment, but available for the free tier before the end of the year!
